File: src/widgets/docker.zig
const std = @import("std");
const tmux_renderer = @import("../tmux_renderer.zig");
const WidgetContext = @import("../core/widget.zig").WidgetContext;

fn getDockerContext(allocator: std.mem.Allocator, io: std.Io) !?[]u8 {
    const result = std.process.run(allocator, io, .{
        .argv = &.{
            "docker", "context", "show",
        },
    }) catch return null;
    defer allocator.free(result.stdout);
    defer allocator.free(result.stderr);

    if (result.term != .exited or result.term.exited != 0) {
        return null;
    }

    const trimmed = std.mem.trim(u8, result.stdout, " \n\r\t");
    if (trimmed.len == 0) {
        return null;
    }

    return try allocator.dupe(u8, trimmed);
}

pub fn run(
    allocator: std.mem.Allocator,
    io: std.Io,
    theme_name: []const u8,
    transparent: bool,
    environ_map: *std.process.Environ.Map,
    writer: *std.Io.Writer,
) !void {
    var ctx = try WidgetContext.init(allocator, io, environ_map, theme_name, transparent);
    defer ctx.deinit();
    const theme = ctx.theme;
    const reset = ctx.reset;

    const context = getDockerContext(allocator, io) catch return;
    defer if (context) |c| allocator.free(c);
    const ctx_str = context orelse return;

    // Color-code: default = muted, else = info
    const color = if (std.mem.eql(u8, ctx_str, "default")) theme.muted else theme.info;

    var hex_buf: [32]u8 = undefined;
    try writer.print("{s}#[fg={s},bg={s},bold] {s}", .{
        reset,
        tmux_renderer.colorHexString(color, &hex_buf),
        tmux_renderer.colorHexString(theme.background, &hex_buf),
        ctx_str,
    });
}

test "docker WidgetContext initializes" {
    const gpa = std.testing.allocator;
    const io = std.Io.threaded_global.ioBasic();
    var env_map = std.process.Environ.Map.init(gpa);
    defer env_map.deinit();

    var ctx = try WidgetContext.init(gpa, io, &env_map, "hard", false);
    defer ctx.deinit();

    try std.testing.expect(!ctx.theme.muted.isDefault());
    try std.testing.expect(!ctx.theme.info.isDefault());
    try std.testing.expect(!ctx.theme.background.isDefault());
    try std.testing.expect(std.mem.startsWith(u8, ctx.reset, "#[fg="));
}

File: src/widgets/docker.rs
use std::process::Command;

use crate::core::cache::{cached_or_compute, scoped_cache_path};
use crate::core::widget::WidgetContext;
use crate::core::Theme;
use crate::tmux_renderer::ThemeHex;

// The active context rarely changes and `docker context show` has real CLI
// startup cost (daemon/socket handshake), so a short cache turns a ~18ms
// subprocess spawn into a near-free file read on most renders. Tradeoff: a
// `docker context use` switch can take up to this long to show up.
const DOCKER_CONTEXT_CACHE_TTL_SECS: u64 = 5;

fn docker_context() -> String {
    match Command::new("docker").args(["context", "show"]).output() {
        Ok(output) if output.status.success() => {
            String::from_utf8(output.stdout).unwrap_or_default().trim().to_owned()
        }
        _ => String::new(),
    }
}

/// Render the Docker context widget.
/// Shows ` <context>` — muted for `default`, info for all others.
pub fn run(theme: Theme) -> String {
    let theme_hex = ThemeHex::from_theme(theme);
    run_with_theme_hex(theme, &theme_hex)
}

pub(crate) fn run_with_theme_hex(theme: Theme, theme_hex: &ThemeHex) -> String {
    let ctx = WidgetContext::from_theme_hex(theme, theme_hex);
    let theme = ctx.theme;

    let cache_path = scoped_cache_path("docker-context", None);
    let context = cached_or_compute(&cache_path, DOCKER_CONTEXT_CACHE_TTL_SECS, docker_context);

    if context.is_empty() {
        return String::new();
    }

    let color = if context == "default" {
        theme.muted
    } else {
        theme.info
    };

    format!(
        "{}#[fg={},bg={},bold] {}",
        ctx.reset,
        theme_hex.color(color),
        theme_hex.color(theme.surface),
        context,
    )
}
